### PR TITLE
PR for frontend backend logic fixes

### LINKED ISSUE #40: Implement Service Detail page
**Branch:** `feature/<issue-number>-service-detail`

### Description
Build the full `ServiceDetailPage` at `/service-detail/:id`. Currently a placeholder. This is the primary page users land on after clicking a service card on the dashboard.

### Acceptance Criteria
- [ ] Fetches `GET /api/services/:id/` on mount; shows a loading skeleton while pending
- [ ] Displays: title, type badge (Offer/Want), description, duration, schedule, location, tags, media gallery
- [ ] Provider card: avatar, name, karma score, featured achievement — links to `/public-profile/:userId`
- [ ] "Request / Offer" CTA button for authenticated users — opens a confirmation modal that calls `POST /api/handshakes/`
- [ ] Comment section: lists comments (`GET /api/services/:id/comments/`) and a reply form for authenticated users
- [ ] Interest counter; "Express Interest" button
- [ ] Share / back navigation
- [ ] Unauthenticated users see a "Log in to request" prompt

### LINKED ISSUE #41: Implement Post Offer and Post Need forms
**Branch:** `feature/<issue-number>-post-service-forms`

### Description
Implement `PostOfferForm` (`/post-offer`) and `PostNeedForm` (`/post-need`). Both are protected routes. They share the majority of their UI and logic; extract a reusable `ServiceForm` component.

### Acceptance Criteria
- [ ] `ServiceForm` component accepts a `type: 'Offer' | 'Need'` prop and renders accordingly
- [ ] Fields: title, description, duration (hours), location type (In-Person / Online), location area (conditional), max participants, schedule type, schedule details, tags (multi-select autocomplete from `GET /api/tags/`), media upload (images)
- [ ] `react-hook-form` + Zod validation; error messages shown per field
- [ ] On submit calls `POST /api/services/`; redirects to the new service detail page on success
- [ ] Tag autocomplete searches existing tags; allows creating new tags inline
- [ ] Media upload with preview thumbnails and remove button
- [ ] "Save as draft" optional stretch goal

---

### LINKED ISSUE #42: Implement Chat & Messaging page
**Branch:** `feature/<issue-number>-chat-messaging`

### Description
Build the real-time messaging interface at `/messages`. Conversations are scoped to a Handshake. Uses the existing `useWebSocket` hook.

### Acceptance Criteria
- [ ] Conversation list on the left: avatar, other user name, last message preview, unread badge; fetched from `GET /api/conversations/`
- [ ] Message thread on the right: chronological messages, sender avatar, timestamp (`timeAgo`); fetched from `GET /api/conversations/:handshakeId/messages/`
- [ ] New message input with send on Enter / send button; calls `POST /api/conversations/:handshakeId/messages/` and emits via WebSocket
- [ ] Incoming messages pushed via WebSocket update the thread in real time without a full re-fetch
- [ ] Selecting a conversation marks messages as read
- [ ] Handshake status banner at the top of the thread (Pending / Accepted / Completed / etc.)
- [ ] Empty state when no conversations exist
- [ ] Responsive layout: on small screens conversation list collapses when a thread is open

### LINKED ISSUE #47: Implement Forum pages
**Branch:** `feature/<issue-number>-forum`

### Description
Implement the four forum pages: `ForumCategories`, `ForumTopicList`, `ForumTopicDetail`, `ForumCreateTopic`.

### Acceptance Criteria

**ForumCategories (`/forum`)**
- [ ] Grid of category cards: icon, name, description, topic count; fetched from `GET /api/forum/categories/`

**ForumTopicList (`/forum/category/:slug`)**
- [ ] List of topics with author, reply count, last activity; fetched from `GET /api/forum/categories/:slug/topics/`
- [ ] "New Topic" button (authenticated only)
- [ ] Pagination

**ForumTopicDetail (`/forum/topic/:topicId`)**
- [ ] Original post + threaded replies; fetched from `GET /api/forum/topics/:id/posts/`
- [ ] Reply form for authenticated users
- [ ] Upvote / downvote on posts
- [ ] Edit / delete own post

**ForumCreateTopic (`/forum/new`)**
- [ ] Category selector, title field, body (markdown or rich text), tag select
- [ ] Calls `POST /api/forum/topics/`; redirects to new topic on success

### PR DESCRIPTION
**Pull request by sgunes16**
Closes #40 #41 #42 #47

---

# PR: Feature Sprint — Service Forms, Chat, Mapbox, Forum

**Branch:** `feature/47-forum` → `dev`  
**Commits:** 21 commits · 55 files changed · +10 851 / −1 054 lines

---

## Table of Contents

1. [Service Post Forms](#1-service-post-forms-closes-41)
2. [Service Detail & Dashboard](#2-service-detail--dashboard-closes-40)
3. [Handshake Workflow Refactor](#3-handshake-workflow-refactor)
4. [Group Chat & Real-time Messaging](#4-group-chat--real-time-messaging-closes-42)
5. [Mapbox Migration & Privacy-Fuzzed Location](#5-mapbox-migration--privacy-fuzzed-location)
6. [Design System Consolidation](#6-design-system-consolidation)
7. [Forum Module](#7-forum-module-closes-47)
8. [Backend Model & Constraint Fixes](#8-backend-model--constraint-fixes)
9. [Test Suite](#9-test-suite)
10. [Infrastructure & Tooling](#10-infrastructure--tooling)
11. [Files Changed](#11-files-changed)
12. [Test Plan](#12-test-plan)

---

### Commit: `ea030d4` — `chore: update Makefile and frontend components for improved service posting`

Created the shared `ServiceForm` component from scratch, wiring it into both `PostOfferForm` and `PostNeedForm` pages.

**`frontend/src/components/ServiceForm.tsx`** (new, 444 lines)
- Built with `react-hook-form` + Zod schema validation.
- Fields: title, description, duration (hours), service type (Offer/Need), location type (In-Person / Online), location area, max participants, schedule type, schedule details, tags, and media upload.
- Tag autocomplete: debounced `GET /api/tags/?search=` with inline create via `POST /api/tags/`.
- Media: `FileReader` API previews, stored as `File[]`, sent as `multipart/form-data`.
- On success: navigates to `/service-detail/:id`.
- Color scheme: Offer → orange/amber, Need → blue/indigo.

**`frontend/src/services/tagAPI.ts`** (new)
- `searchTags(query)` — debounced autocomplete.
- `createTag(name)` — inline tag creation.

**`frontend/src/pages/PostOfferForm.tsx` / `PostNeedForm.tsx`**
- Reduced to thin wrappers that pass `type="Offer"` or `type="Need"` to `ServiceForm`.

**`frontend/src/components/ProtectedRoute.tsx`**
- Refactored; onboarding redirect disabled (backend does not yet expose `is_onboarded`).

---

### Commit: `29885ce` — `fix: fix ServiceForm component structure and validation and lint`

- Removed redundant `Box` wrapper around `<form>`.
- Tightened Zod resolver type to satisfy TypeScript strict mode.
- Cleaned up unused schema exports and collection helpers.

---

### Commit: `f61ed61` — `feat: implement service and handshake management features`

**`frontend/src/pages/ServiceDetailPage.tsx`** (grew from stub → 933 lines)
- Full service detail view: title, type badge, duration, schedule, location, tags, description.
- `usePolling`-powered refresh for live handshake status.
- "Express Interest" button → creates a handshake; button state reflects current handshake status (`Pending`, `Accepted`, `Agreed`, `Completed`, `Cancelled`).
- Handshake status badge map with distinct colours per state.
- Comments/reviews section with paginated list and post form.
- Navigate to `/messages/:handshakeId` from detail page once handshake exists.

**`frontend/src/pages/DashboardPage.tsx`** (major expansion, +298 lines)
- Service list with polling for live updates.
- Filter bar: type toggle, search, distance slider.
- `usePolling` hook replaces manual `setInterval`.

**`frontend/src/services/commentAPI.ts`** (new, 71 lines)
- `list(serviceId)`, `create(serviceId, body)`, `update(id, body)`, `delete(id)`.

**`frontend/src/services/handshakeAPI.ts`** (new, 38 lines)
- `list()`, `accept(id)`, `cancel(id)`, `confirm(id)`.

**`frontend/src/services/serviceAPI.ts`**
- Added `page` and `tags` filter params to `list()`.
- Added `interest(id)` to express interest and create a handshake.

---

### Commit: `84f70c6` — `refactor: improve handshake service logic for one-time services`

**`backend/api/services.py`**
- `capacity_reached()` now excludes `'pending'` handshakes from the count — pending requests never block new interest expressions.
- `accept()` auto-denies all other pending handshakes when accepting on a one-time service; sends denial notifications.
- `existing_interest_status()` prevents a user from expressing interest twice.

**`backend/api/views.py`**
- `InterestView` checks capacity before creating a handshake.
- Returns correct HTTP 409 when service is full.

---

### Commit: `85a3f56` — `feat: add 'Agreed' status and participant count to services`

**`backend/api/models.py`**
- Added `'Agreed'` to `Service.STATUS_CHOICES` — sits between `Active` and `Completed`.

**`backend/api/serializers.py`**
- `participant_count` — `SerializerMethodField` counting only `Accepted` + `Completed` handshakes.

**`backend/api/services.py` / `views.py`**
- When accepted participant count reaches `max_participants`, service transitions to `Agreed` automatically.
- All remaining pending handshakes are denied with notifications.

**Migration:** `0028_add_agreed_service_status`

---

### Commit: `1e59f67` — `feat: enhance handshake workflow for one-time services for demo`

**`backend/setup_demo.py`**
- Demo script back-dates handshake timestamps.
- Auto-transitions services to `Completed` when no active handshakes remain.
- Mirrors production status logic for consistent demo data.

---

### Commit: `4da9f2d` — `feat: refine handshake workflow for one-time services`

**`backend/api/views.py`**
- `Agreed` transition only fires once all slots are genuinely filled (idempotency guard).
- Denial notifications use a dedicated notification type with service title context.

**`backend/setup_demo.py`**
- Demo seeding covers the full `Active → Agreed → Completed` lifecycle.

---

### Commit: `9f879c8` — `feat: update handshake initiation and approval logic for service owner model`

Clarified the role split in the handshake lifecycle:

| Action | Who can perform it |
|---|---|
| `POST /handshakes/:id/initiate/` | **Service owner only** (proposes meeting details) |
| `POST /handshakes/:id/accept/` | **Requester only** (approves the proposal) |

Works identically for both Offer and Need service types. Enforced via `IsServiceOwner` / `IsRequester` permission classes in `views.py`.

---

### Commit: `533f841` — `feat: add group chat functionality for private services`

**`backend/api/models.py`** — new `ServiceGroupChatMessage` model:
```
ServiceGroupChatMessage
├── id          UUIDField (PK)
├── service     FK → Service
├── sender      FK → User
├── body        TextField
└── created_at  auto
Indexes: (service, created_at), (sender)
```

**`backend/api/consumers.py`** — `GroupChatConsumer` (WebSocket):
- Room name: `service_group_<uuid>`.
- On connect: verifies caller has an `Accepted` handshake for the service; disconnects otherwise.
- Broadcasts messages to all room participants.

**`backend/api/views.py`** — `GroupChatViewSet`:
- `GET /api/group-chat/<service_uuid>/` — paginated message history.
- `POST /api/group-chat/<service_uuid>/` — send message.
- Returns 403 if caller has no accepted handshake.

**`backend/api/urls.py` / `routing.py`**
- REST endpoint and WebSocket route wired.

**Migration:** `0029_add_service_group_chat_message`

---

### Commit: `f757d25` — `chore: update backend server command in Makefile`

- Changed `runserver` → `daphne` in `make dev` to enable ASGI/WebSocket support in development.

---

### Commit: `183f831` — `feat: add usePolling hook for generic polling functionality`

**`frontend/src/hooks/usePolling.ts`** (new, 121 lines):
- Generic hook: `usePolling(asyncFn, { interval, immediate, pauseOnHidden })`.
- Manages `loading`, `error`, and `data` state.
- Pauses polling when the browser tab is hidden (`visibilitychange`).
- Exposes `refresh()` for manual trigger.
- Used by `ChatPage` as WebSocket fallback and by `ServiceDetailPage`/`DashboardPage` for live updates.

---

### Commit: `fb33c8c` — `fix: update WebSocket target and refine handshake API`

**`frontend/vite.config.ts`**
- WebSocket proxy target fixed to use `ws://localhost:8000` directly (not the Vite dev origin).

**`frontend/src/hooks/useWebSocket.ts`**
- Distinguishes permanent rejections (4001 = unauthorized, 4003 = forbidden) from transient failures.
- Permanent rejection → stops reconnect loop immediately.

**`frontend/src/services/handshakeAPI.ts`**
- Added `initiate(id, payload)` — posts meeting proposal `{ proposed_time, location, notes }`.
- Added `approve(id)` — requester approves the proposal.

---

### Commit: `5e460e4` — `feat: enhance chat functionality with new modals and routing`

**`frontend/src/pages/ChatPage.tsx`** (full rewrite, 1 650+ lines):
- Two-panel layout: conversation list (left) + message thread (right).
- Supports both 1-to-1 handshake chat and new service **group chat** tab.
- Real-time via `useWebSocket`; graceful fallback to `usePolling` on disconnect.
- Message bubbles: sent → right (green), received → left (white).
- Auto-scroll to latest; unread count badge in conversation list.
- Inline handshake action bar (accept / initiate / confirm / cancel) with status-aware rendering.

**New components:**

| Component | Purpose |
|---|---|
| `HandshakeDetailsModal.tsx` (159 lines) | Shows handshake state, proposed meeting details, action buttons (initiate / approve) |
| `ProviderDetailsModal.tsx` (136 lines) | Public profile of the other party, avatar, rating, completed exchange count |
| `ServiceConfirmationModal.tsx` (112 lines) | Two-step confirm/cancel for service completion with reason field |

**`frontend/src/services/conversationAPI.ts`** (new, 134 lines):
- `listConversations()` — chat sidebar data.
- `getMessages(handshakeId, page)` — paginated message history.
- `sendMessage(handshakeId, body)` — REST fallback when WebSocket is down.

**`frontend/src/App.tsx`**
- `/messages` and `/messages/:handshakeId` routes added; lazy-loaded.

**`frontend/src/components/Navbar.tsx`**
- Messages link and unread badge added to nav items.

---

### Commit: `915a094` — `feat: implement ServiceGroupChatMessage related tests`

- `ServiceGroupChatMessageFactory` added to `helpers/factories.py`.
- Integration tests (`test_chat_api.py` +265 lines): access control, message retrieval for owner and accepted participants, pagination.
- Integration tests (`test_service_api.py` +94 lines): service capacity edge cases via group chat context.
- Unit tests (`test_group_chat.py` 228 lines): model field validation, serializer output, consumer auth logic.

---

### Commit: `66643a2` — `feat: integrate Mapbox for enhanced mapping functionality`

Replaced `react-leaflet` with `react-map-gl` (Mapbox GL JS).

**`frontend/src/components/MapView.tsx`** (rewritten, +731 lines):
- Vector tile base map via `react-map-gl/mapbox`.
- **Clustering** — GeoJSON source with `cluster: true`; clicking a cluster flies in one zoom level.
- **Custom markers** — Offer (amber), Need (indigo) colour-coded circle layers.
- **2 km fuzzy circle** — semi-transparent `circle` layer centred on `circle_lat/lng` (backend-provided); inner dot hidden so raw location is never visually revealed.
- **Popup** — anchored to the **right** of the marker (`anchor="left"`, `offset=16`); shows title, type, duration, author avatar — never raw coordinates.
- **User location** — blue dot + animated halo layer.
- **"My Location" button** (bottom-right) — `flyTo` user coords with 1 200 ms ease.
- **Auto-centre** — opens on user geolocation if granted; falls back to single In-Person service's fuzzed coordinates.
- **Stack jitter** — deterministic ~55 m per-service offset prevents identical coordinates from stacking into the same pixel.
- Cluster count labels and cluster circle removed — user zooms in manually.

**`frontend/src/components/ServiceForm.tsx`** (major additions):
- **Mapbox Geocoding address search** — `LocationSearch` sub-component with debounced autocomplete dropdown.
- **"Use My Location" button** — reverse-geocodes browser position and populates the address field.
- `extractDistrict()` helper — stores only the *ilçe* (district) from Mapbox `context[]` as `location_area` (≤ 100 chars).
- Coordinates rounded to `toFixed(6)` before API submission.
- DRF field-level errors parsed and shown per-field.

**`frontend/src/pages/ServiceDetailPage.tsx`**
- Mini-map (220 px height) for In-Person services with 2 km privacy circle.
- Location info tile shows district name only.
- Privacy note: *"Exact address is hidden — shown within a 2 km privacy zone."*

**`frontend/.env.example`**
- `VITE_MAPBOX_TOKEN=` documented.

**`DEPLOYMENT.md`**
- Mapbox token setup instructions added.

**`frontend/THEME.md`** (new, 303 lines)
- Full design system documentation: colour tokens, component patterns, typography, spacing, animation guidelines.

---

### Commit: `f1dae80` — `refactor: optimize MapView component for improved user location handling`

- `useMemo` hooks for `userGeoJSON` and `initialView` moved **before** the `if (!TOKEN)` early return — fixes React Rules of Hooks violation.
- `initialView` logic: centres on a single visible In-Person service when user location is not yet available.

---

### Commit: `0698c6b` — `feat: implement privacy-fuzzed coordinates for in-person services`

**Backend — `serializers.py`:**

```python
def _fuzzy_coords(lat, lng, seed):
    # FNV-1a deterministic hash → pseudo-random angle + ~1 km radius offset
    h = fnv1a(seed)
    angle = (h & 0xFFFF) / 0xFFFF * 2 * math.pi
    r = 0.009  # ≈ 1 km in degrees
    return lat + r * math.sin(angle), lng + r * math.cos(angle)
```

`ServiceSerializer.to_representation` now returns:

| JSON field | Value |
|---|---|
| `location_lat` / `location_lng` | DB coords + `_fuzzy_coords(seed="<id>_m")` ≈ 1 km offset (marker position) |
| `circle_lat` / `circle_lng` | DB coords + `_fuzzy_coords(seed="<id>_c")` ≈ 1 km offset (circle centre) |

The two offsets use different seeds, so the marker and circle centre land in **different** random directions. Neither field reveals the real coordinate. Raw DB values are never serialized.

**Frontend — `MapView.tsx`:**
- `idFuzzyOffset` function removed — frontend no longer needs to apply its own offset.
- `buildGeoJSON` uses `location_lat/lng` directly (already fuzzed by backend).
- `buildCircleGeoJSON` uses `circle_lat/lng` directly.
- Stack jitter reduced from 500 m to ~55 m (visual separation only; privacy handled by backend).

**`DashboardPage.tsx`:**
- When location is active, two parallel `serviceAPI.list` calls are made via `Promise.all`:
  1. Nearby In-Person services (with `lat`, `lng`, `distance` params).
  2. All Online services (no distance filter).
- Results are merged and deduplicated by `id`.

---

### Commit: `c4abb9e` — `refactor: consolidate design tokens and improve component styling`

**`frontend/src/theme/tokens.ts`** (new, 38 lines):

```typescript
// Brand
export const GREEN    = '#2D5C4E'
export const GREEN_LT = '#F0FDF4'
export const YELLOW   = '#F8C84A'

// Accent
export const AMBER    = '#D97706'   export const AMBER_LT = '#FFFBEB'
export const BLUE     = '#1D4ED8'   export const BLUE_LT  = '#EFF6FF'
export const PURPLE   = '#7C3AED'   export const PURPLE_LT = '#F3E8FF'

// Semantic
export const RED      = '#DC2626'   export const RED_LT   = '#FEF2F2'

// Neutrals
export const GRAY50 … GRAY900, WHITE
```

All pages and components updated to import from `tokens.ts` — no more inline hex literals scattered across files.

**`frontend/src/components/Navbar.tsx`** (592 lines refactored):
- Solid white background (removed `backdropFilter` blur that caused DPI/antialiasing issues).
- `WebkitFontSmoothing: antialiased` for crisp text.
- Responsive mobile menu with token-based colours.
- Messages icon with unread badge.

**`frontend/src/pages/DashboardPage.tsx`** (major layout overhaul):
- Two-panel card design matching ChatPage aesthetic.
- Map panel open by default.
- Service cards redesigned: `MetaChip` components for duration/type/schedule, provider row, hover states.
- Distance slider always visible when location is permitted.
- `flyTo` fires only once on first location grant (guarded by `didFlyToUser` ref) — eliminates map "shift" on page refresh.

**`frontend/src/components/ServiceForm.tsx`** / **`PostNeedForm.tsx`** / **`PostOfferForm.tsx`**:
- Token imports replace all hardcoded colours.
- Layout improvements: cleaner section spacing, better mobile stack behaviour.

---

#### `frontend/src/services/forumAPI.ts` (new, 67 lines)

Full CRUD wrapper using the shared `apiClient`:

```typescript
forumAPI.listCategories()           // GET /forum/categories/
forumAPI.getCategory(slug)          // GET /forum/categories/:slug/
forumAPI.listTopics({ category, page, page_size })  // GET /forum/topics/
forumAPI.getTopic(id)               // GET /forum/topics/:id/
forumAPI.createTopic({ title, body, category })     // POST /forum/topics/
forumAPI.listPosts(topicId, { page, page_size })    // GET /forum/topics/:id/posts/
forumAPI.createPost(topicId, body)  // POST /forum/topics/:id/posts/
forumAPI.updatePost(postId, body)   // PATCH /forum/posts/:id/
forumAPI.deletePost(postId)         // DELETE /forum/posts/:id/
```

> **Note:** `POST /api/forum/topics/` accepts `category` as a **slug** string, not a UUID.

---

#### `/forum` — `ForumCategories.tsx`

- Grid (1 / 2 / 3 columns responsive) of category cards sorted by `display_order`.
- Each card: 4 px coloured top strip, icon (string → `react-icons/fi` map), name, topic count, description (2-line clamp), post count, last activity.
- Icon strings supported: `message-square`, `book-open`, `calendar`, `users`, `star`, `lightbulb`, `globe`, `code`, `heart`, `home`, `tool`, `award`.
- Colour map: `blue`, `green`, `purple`, `amber`, `orange`, `pink`, `red`, `teal`.
- Loading skeleton (pulse animation) + empty state + error state.
- Click → `navigate(/forum/category/:slug)`.

---

#### `/forum/category/:slug` — `ForumTopicList.tsx`

- Fetches category metadata and topic list in parallel (`Promise.all`).
- Pinned topics float to the top of the list.
- Each row: author avatar (deterministic colour from name initial), pin / lock badge, title, author name, last activity, reply count, view count.
- **"New Topic"** button — visible to authenticated users only; pre-fills `?category=<slug>`.
- Pagination: 15 topics per page; ellipsis-aware page number strip.
- Empty state with CTA for authenticated users.

---

#### `/forum/topic/:topicId` — `ForumTopicDetail.tsx`

- Fetches topic metadata and first page of posts in parallel.
- **OP block**: green (`GREEN_LT`) background, "Author" green badge, full body text.
- **Reply list**: paginated (20 per page); "OP" badge on any reply from the topic author.
- **Edit own post**: inline textarea replaces body; "Save" / "Cancel" buttons.
- **Delete own post**: two-step confirmation (icon → confirm/cancel row) before API call; soft-delete shows `[This post was deleted]`.
- **Reply form** (authenticated): auto-appends to current page if room; otherwise navigates to final page.
- **Locked topic**: reply box replaced by lock-icon banner.
- **Unauthenticated**: "Sign in to join the discussion" prompt with link.
- Pagination for replies with prev/next buttons.

---

#### `/forum/new` — `ForumCreateTopic.tsx`

- `?category=<slug>` query param pre-selects the category `<select>`.
- `react-hook-form` + Zod schema:
  - `category`: required.
  - `title`: 5–200 characters.
  - `body`: 20–10 000 characters.
- Character counter on body field; counter turns red at 10 000.
- Tips panel (green) with community guidelines.
- On success: `navigate(/forum/topic/:id)` to the new topic.
- On error: DRF `detail` message shown via `toast.error`.

---

### Commit: `8161fa7` — `refactor: update CheckConstraint syntax in models and enhance UserFactory password handling`

**`backend/api/models.py`** — Django 5.x compatibility:

All `CheckConstraint(check=...)` updated to `CheckConstraint(condition=...)`:

| Model | Constraint |
|---|---|
| `User` | `timebank_balance >= -10.00` |
| `Service` | `duration > 0`, `max_participants > 0` |
| `Handshake` | `provisioned_hours > 0` |

**`backend/api/tests/helpers/factories.py`**:
- `UserFactory` uses `@post_generation` hook to set hashed passwords, fixing silent auth failures in tests.

---

### Commit: `5d7faee` — `feat: add comprehensive tests for capacity management and service status transitions`

| File | Lines | Coverage |
|---|---|---|
| `tests/unit/test_capacity_and_agreed.py` | 511 | Capacity limits, `Agreed` auto-transition, pending-not-blocked rule, denial notifications |
| `tests/unit/test_serializers.py` | 133 | `participant_count` correctness (Accepted + Completed only), fuzzy coord determinism |
| `tests/unit/test_property_tests.py` | +39 | Timebank balance invariants, service participation upper bounds |
| `tests/integration/test_handshake_api.py` | +238 | Full accept → deny-others → Agreed lifecycle; edge cases |
| `tests/unit/test_services.py` | +30 | Updated for new handshake acceptance logic |

### Commit: `9f879c8` (integration tests)

`tests/integration/test_handshake_api.py` +189 lines:
- `initiate` permission: only service owner may call it.
- `accept/approve` permission: only requester may call it.
- Both permissions tested for Offer and Need service types.
- Correct 403 returned for wrong-role attempts.

### Commit: `915a094` (group chat tests)

| File | Lines | Coverage |
|---|---|---|
| `tests/unit/test_group_chat.py` | 228 | Model fields, serializer output, consumer auth |
| `tests/integration/test_chat_api.py` | +265 | Access control, CRUD, pagination for group chat |
| `tests/integration/test_service_api.py` | +94 | Capacity edge cases exposed via service API |

---

### Makefile changes across commits
- `ea030d4`: `BACKEND_URL=http://localhost:8000` exported for local dev.
- `f757d25`: Dev backend command changed from `python manage.py runserver` → `daphne hive_project.asgi:application` for ASGI/WebSocket support.

### `DEPLOYMENT.md`
- Mapbox token (`VITE_MAPBOX_TOKEN`) setup steps added.
- SSL / Certbot section already present from Agent-1; not modified.

### `frontend/.env.example`
```
VITE_API_URL=/api
VITE_MAPBOX_TOKEN=pk.your_token_here
```

---

### Backend (23 files)

| File | Change |
|---|---|
| `api/models.py` | `Agreed` status, `ServiceGroupChatMessage` model, Django 5.x constraint fix |
| `api/serializers.py` | `participant_count`, fuzzy coords (`location_lat/lng`, `circle_lat/lng`), group chat serializer |
| `api/views.py` | `GroupChatViewSet`, `InterestView` capacity check, handshake permission split, `Agreed` transition |
| `api/services.py` | Capacity logic, auto-deny, owner-model handshake |
| `api/consumers.py` | `GroupChatConsumer` (WebSocket) |
| `api/routing.py` | Group chat WS route |
| `api/urls.py` | `/group-chat/<uuid>/` REST route |
| `api/utils.py` | Minor utility update |
| `api/migrations/0028_*` | `Agreed` status |
| `api/migrations/0029_*` | `ServiceGroupChatMessage` table + indexes |
| `api/tests/helpers/factories.py` | `UserFactory` password fix, `ServiceGroupChatMessageFactory` |
| `api/tests/unit/test_capacity_and_agreed.py` | New (511 lines) |
| `api/tests/unit/test_group_chat.py` | New (228 lines) |
| `api/tests/unit/test_serializers.py` | New (133 lines) |
| `api/tests/unit/test_property_tests.py` | Extended (+39 lines) |
| `api/tests/unit/test_services.py` | Updated (+30 lines) |
| `api/tests/integration/test_handshake_api.py` | Extended (+427 lines) |
| `api/tests/integration/test_chat_api.py` | Extended (+265 lines) |
| `api/tests/integration/test_service_api.py` | Extended (+94 lines) |
| `setup_demo.py` | Full lifecycle demo seeding |
| `Makefile` | Daphne dev server |
| `DEPLOYMENT.md` | Mapbox token docs |

### Frontend (32 files)

| File | Change |
|---|---|
| `src/theme/tokens.ts` | New — centralized colour tokens |
| `src/types/index.ts` | `ForumCategory`, `ForumTopic`, `ForumPost` types |
| `src/App.tsx` | `/messages/:handshakeId`, forum routes wired |
| `vite.config.ts` | WebSocket proxy target fixed |
| `src/hooks/usePolling.ts` | New (121 lines) |
| `src/hooks/useWebSocket.ts` | Permanent vs transient rejection logic |
| `src/services/tagAPI.ts` | New |
| `src/services/commentAPI.ts` | New (71 lines) |
| `src/services/handshakeAPI.ts` | New + extended (83 lines) |
| `src/services/conversationAPI.ts` | New (134 lines) |
| `src/services/serviceAPI.ts` | Pagination, tags filter, `interest()` |
| `src/services/forumAPI.ts` | New (67 lines) |
| `src/components/ServiceForm.tsx` | Complete build (833 lines) |
| `src/components/MapView.tsx` | Full Mapbox rewrite (627 lines) |
| `src/components/Navbar.tsx` | DPI fix, messages badge, token colours |
| `src/components/ProtectedRoute.tsx` | Onboarding redirect disabled |
| `src/components/HandshakeDetailsModal.tsx` | New (159 lines) |
| `src/components/ProviderDetailsModal.tsx` | New (136 lines) |
| `src/components/ServiceConfirmationModal.tsx` | New (112 lines) |
| `src/pages/DashboardPage.tsx` | Major redesign (983 lines) |
| `src/pages/ServiceDetailPage.tsx` | Full implementation (976 lines) |
| `src/pages/ChatPage.tsx` | Full rewrite (1 650 lines) |
| `src/pages/PostOfferForm.tsx` | Thin wrapper |
| `src/pages/PostNeedForm.tsx` | Thin wrapper |
| `src/pages/ForumCategories.tsx` | New (207 lines) |
| `src/pages/ForumTopicList.tsx` | New (323 lines) |
| `src/pages/ForumTopicDetail.tsx` | New (508 lines) |
| `src/pages/ForumCreateTopic.tsx` | New (265 lines) |
| `frontend/THEME.md` | New (303 lines) |
| `frontend/.env.example` | Mapbox token |
| `package.json` / `package-lock.json` | `react-map-gl`, `mapbox-gl`, `@types/geojson`, `zod`, `@hookform/resolvers` |

---

### Backend
```bash
cd backend
python manage.py test api.tests.unit.test_capacity_and_agreed
python manage.py test api.tests.unit.test_group_chat
python manage.py test api.tests.unit.test_serializers
python manage.py test api.tests.integration.test_handshake_api
python manage.py test api.tests.integration.test_chat_api
python manage.py test api.tests.integration.test_service_api
python manage.py test   # full suite
```

### Frontend
```bash
cd frontend
npm run build           # must exit 0, 0 TypeScript errors
```

### Manual Checklist

**Service Forms**
- [ ] Post an Offer at `/post-offer` — tag autocomplete works, "Use My Location" populates address field
- [ ] Post a Need at `/post-need` — form validates correctly, submits, redirects to detail
- [ ] Submit with invalid data — per-field Zod errors shown

**Service Detail & Handshake**
- [ ] `/service-detail/:id` loads with full info, comments, and interest button
- [ ] Express interest → handshake created; button reflects `Pending` state
- [ ] Accept handshake (as service owner) → status changes to `Accepted`; other pending requests auto-denied
- [ ] Fill all slots → service status transitions to `Agreed`
- [ ] Confirm completion → status becomes `Completed`

**Chat & Group Chat**
- [ ] `/messages` lists all conversations
- [ ] Clicking a conversation opens the thread; WebSocket receives new messages in real-time
- [ ] Group chat tab visible for multi-participant services; only accepted participants can message
- [ ] Disconnect network briefly → polling fallback activates; reconnect → WebSocket resumes

**Map & Location**
- [ ] Dashboard map opens on user's location (if permitted); distance slider visible
- [ ] "My Location" button re-centres map with smooth animation
- [ ] Offer markers (amber) and Need markers (indigo) render correctly
- [ ] Clicking a marker opens a right-anchored popup; no raw coordinates shown
- [ ] 2 km semi-transparent circle visible around each In-Person service
- [ ] `location_lat/lng` and `circle_lat/lng` in API response are fuzzed and distinct from each other
- [ ] Service Detail mini-map shows privacy circle; location tile shows district name only

**Forum**
- [ ] `/forum` — category grid loads with correct icons and colours
- [ ] `/forum/category/general` — topic list with reply/view counts; "New Topic" hidden when logged out
- [ ] Create a topic at `/forum/new?category=general` → redirects to new topic
- [ ] Post a reply → appears at bottom of thread
- [ ] Edit own reply inline; save and cancel work
- [ ] Delete own reply → shows `[This post was deleted]`
- [ ] Locked topic → reply form replaced by lock banner
- [ ] Pagination works on both topic list and topic detail

### CODE REVIEW

**Comment by sgunes16** [comment]
@yaseminsirin you can use test plan from here

File: backend/api/views.py [hunk 1]
```diff
@@ -798,6 +814,34 @@ def is_valid_coordinate(value: str | None) -> bool:
...
+            .prefetch_related(
+                'tags',
+                user_badges_prefetch,
+                Prefetch('media', queryset=ServiceMedia.objects.order_by('display_order', 'created_at')),
+                capacity_handshakes_prefetch,
+            )
+        )
+        instance = get_object_or_404(queryset, pk=kwargs['pk'])
+        serializer = self.get_serializer(instance)
+        return Response(serializer.data)
```
[the remark this answers is not in the document]
Comment (sgunes16) [reply backend/api/views.py:843]: I did my own. I want to link services to closed chats. So, hiding the services from non owner users is unnecessary. There is no secret info at the services. Also people cannot interest again if the service is completed.

File: frontend/src/services/serviceAPI.ts [hunk 2]
```diff
@@ -8,6 +8,9 @@ export interface ServiceListParams {
   search?: string
   type?: 'Offer' | 'Need'
   status?: string
+  tags?: string
+  page?: number
+  page_size?: number
```
[the remark this answers is not in the document]
Comment (sgunes16) [reply frontend/src/services/serviceAPI.ts:13]: nice catch

File: frontend/src/services/conversationAPI.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    const res = await apiClient.post<GroupChatMessage>(`/group-chat/${serviceId}/`, { body })
+    return res.data
+  },
+}
+
+// Dev  → direct to Daphne (Vite's WS proxy intercepts HMR and drops /ws upgrades)
+// Prod → same host (Nginx handles the upgrade)
+const wsBase = import.meta.env.DEV
+  ? 'ws://localhost:8000'
+  : `${window.location.protocol === 'https:' ? 'wss' : 'ws'}://${window.location.host}`
```
[the remark this answers is not in the document]
Comment (sgunes16) [reply frontend/src/services/conversationAPI.ts]: for docker compose dev, it might throw error. It should have tested

File: backend/api/serializers.py [hunk 4]
```diff
@@ -238,6 +239,26 @@ def validate_file_url(self, value):
...
+
+    1 km ≈ 0.009° latitude at Istanbul's latitude (~41°N).
+    """
+    h = 2166136261
+    for ch in service_id:
+        h ^= ord(ch)
+        h = (h * 16777619) & 0xFFFFFFFF   # unsigned 32-bit, matches JS Math.imul + >>> 0
+
+    angle = (h / 0xFFFFFFFF) * 2 * math.pi
+    R = 0.0045   # ~1 km
```
[the remark this answers is not in the document]
Comment (sgunes16) [reply backend/api/serializers.py]: I changed to 0.045 intentionally. Comment is wrong. mb

**Review by mzyavuz — DISMISSED** [review]
Wikidata parts are missing, I will add them now.

**Review by mzyavuz — APPROVED** [review]
Great work
